pathway-analysis/02-gene_set_enrichment_analysis.Rmd

```diff
@@ -111,77 +126,121 @@ dge_results_df <- readr::read_tsv(dge_results_file)
+The function `sum()` will count `TRUE` values as 1s and `FALSE` as 0s, so using it with `duplicated()` will count the number of duplicate values.
+
+```{r any_duplicated, live = TRUE}
+sum(duplicated(dge_results_df$gene_symbol))
+```
+
+This will cause a problem when we go to run GSEA.
+
+### Removing duplicates
+
+The GSEA approach requires on discriminating between genes that are in a gene set and those that are not. 
+Practically speaking, gene sets are just collections of gene identifiers!
+When the function we use for GSEA pre-ranked gets a list with duplicated gene identifiers, it can produce unexpected results.
+
+Compared to the total number of genes that are in our results, there are not a lot of duplicates but we'll still need to make a decision about how to handle them.
+
+Let's get a vector of the duplicated gene symbols so we can use it to explore our filtering steps.
+
+```{r gene_dups, live = TRUE}
+duplicated_gene_symbols <- dge_results_df %>%
+  dplyr::filter(duplicated(gene_symbol)) %>%
+  dplyr::pull(gene_symbol)
+```
 
-When `mapIds()` is run with `multiValues = "first"` and it encounters multiple matches, only the first identifier is returned.
-This is the default behavior.
-It will also return a named vector of the IDs you queried for (`column`), where the names are the input identifiers (`keys`), in the same order as the keys which means you can use it directly with `dplyr::mutate()`.
+Now we'll look at the values for the the duplicated gene symbols.
 
-```{r convert_entrez, live = TRUE}
-dge_results_df <- dge_results_df %>%
-  # Create a new column 'entrez_id' that contains the Entrez IDs returned by
-  # mapIds()
-  dplyr::mutate(entrez_id = mapIds(org.Hs.eg.db, 
-                                   keys = dge_results_df$Gene, 
-                                   column = "ENTREZID", 
-                                   keytype = "ENSEMBL", 
-                                   multiVals = "first")
-  ) %>%
-  # We need these gene identifiers to perform our pathway analysis, so remove
-  # any genes where we don't have Entrez IDs
-  dplyr::filter(!is.na(entrez_id))
+```{r show_gene_dups}
+dge_results_df %>%
+  dplyr::filter(gene_symbol %in% duplicated_gene_symbols) %>%
+  dplyr::arrange(gene_symbol)
 ```
 
-### Preranked list
+We can see that the associated values vary for each row.
 
-The `GSEA()` function takes a preranked (sorted) named vector of statistics, where the names in the vector are gene identifiers.
+Let's keep the gene symbols associated with the higher absolute value of the log2 fold change.
+GSEA relies on genes' rankings on the basis of a gene-level statistic and the enrichment score that is calculated reflects the degree to which genes in a gene set are overrepresented in the top or bottom of the rankings ([Subramanian _et al._ 2005](https://doi.org/10.1073/pnas.0506580102); [Yu](http://yulab-smu.top/clusterProfiler-book/chapter2.html#gene-set-enrichment-analysis)).
```

Comment:
I think we could lose this sentence, given the earlier explanation. (The rest of the explanation seems good to me.)

```suggestion
```